File: frontend/src/utils/agreementIndicators.js
export const REQUESTER_ENTITY_KIND_OPTIONS = [
  { value: "SCHOOL", label: "Instituição de Ensino" },
  { value: "BUSINESS", label: "Empresa" },
  { value: "MILITARY", label: "Órgão Militar" },
  { value: "PUBLIC", label: "Órgão Público" },
  { value: "OTHER", label: "Outros" },
];

export const REQUESTER_ENTITY_NATURE_OPTIONS = [
  { value: "PUBLIC", label: "Pública" },
  { value: "PRIVATE", label: "Privada" },
];

export const EDUCATION_ACTION_AGE_RANGE_OPTIONS = [
  { value: "AGE_05_10", label: "05 - 10 anos (ensino fundamental - anos iniciais)" },
  { value: "AGE_11_14", label: "11 - 14 anos (ensino fundamental - anos finais)" },
  { value: "AGE_15_17", label: "15 - 17 anos (ensino médio)" },
  { value: "AGE_ADULT", label: "acima de 18 anos - Adultos" },
];

const AGE_LABEL_TO_VALUE = new Map(
  EDUCATION_ACTION_AGE_RANGE_OPTIONS.map((option) => [option.label.toLocaleLowerCase("pt-BR"), option.value])
);

function normalizedText(value) {
  return String(value || "").trim().toLocaleLowerCase("pt-BR");
}

export function agreementIndicatorLabel(value) {
  return value === "ESCOLINHA_NOTA_10"
    ? "Escolinha Nota 10"
    : value === "ESCOLA_NOTA_10"
      ? "Escola Nota 10"
      : "";
}

export function requesterEntityKindLabel(value) {
  return REQUESTER_ENTITY_KIND_OPTIONS.find((option) => option.value === value)?.label || value || "";
}

export function requesterEntityNatureLabel(value) {
  return REQUESTER_ENTITY_NATURE_OPTIONS.find((option) => option.value === value)?.label || value || "";
}

export function ageRangeLabel(value) {
  return EDUCATION_ACTION_AGE_RANGE_OPTIONS.find((option) => option.value === value)?.label || value || "";
}

export function normalizeAgeRange(value) {
  const text = String(value || "").trim();
  if (!text) return "";
  if (EDUCATION_ACTION_AGE_RANGE_OPTIONS.some((option) => option.value === text)) return text;
  return AGE_LABEL_TO_VALUE.get(normalizedText(text)) || "";
}

export function normalizeRequesterEntityType(value) {
  const text = String(value || "").trim();
  if (!text) return { kind: "", nature: "" };
  if (REQUESTER_ENTITY_KIND_OPTIONS.some((option) => option.value === text)) return { kind: text, nature: "" };

  const lower = normalizedText(text);
  if (text === "2" || lower.includes("instituição de ensino") || lower.includes("instituicao de ensino") || lower.includes("escola") || lower.includes("colégio") || lower.includes("colegio")) {
    const nature = lower.includes("públic") || lower.includes("public") ? "PUBLIC" : lower.includes("privad") || lower.includes("particular") ? "PRIVATE" : "";
    return { kind: "SCHOOL", nature };
  }
  if (text === "4" || lower.includes("empresa")) return { kind: "BUSINESS", nature: "" };
  if (lower.includes("militar")) return { kind: "MILITARY", nature: "" };
  if (text === "1" || lower.includes("órgão") || lower.includes("orgao")) return { kind: "PUBLIC", nature: "" };
  return { kind: "", nature: "" };
}

export function deriveEducationAgreementIndicator(kind, nature, ageRange) {
  const normalizedAge = normalizeAgeRange(ageRange);
  if (kind !== "SCHOOL" || nature !== "PUBLIC" || !normalizedAge) return "";
  if (normalizedAge === "AGE_05_10" || normalizedAge === "AGE_11_14") return "ESCOLINHA_NOTA_10";
  if (normalizedAge === "AGE_15_17" || normalizedAge === "AGE_ADULT") return "ESCOLA_NOTA_10";
  return "";
}

export function deriveAgreementIndicatorFromAgenda(agenda) {
  const fields = buildAgreementFieldsFromAgenda(agenda);
  return deriveEducationAgreementIndicator(
    fields.requester_entity_kind,
    fields.requester_entity_nature,
    fields.age_range
  );
}

export function buildAgreementFieldsFromAgenda(agenda) {
  const mappedType = normalizeRequesterEntityType(agenda?.requester_entity_type);
  const directKind = normalizeRequesterEntityType(agenda?.requester_entity_kind).kind;
  const directNature = String(agenda?.requester_entity_nature || "").trim();
  return {
    // External requests keep their description in requester_entity_type, while
    // internal requests may retain the select values in separate fields.
    requester_entity_kind: directKind || mappedType.kind,
    requester_entity_nature: REQUESTER_ENTITY_NATURE_OPTIONS.some((option) => option.value === directNature)
      ? directNature
      : mappedType.nature,
    age_range: normalizeAgeRange(agenda?.age_range || agenda?.age_ranges),
  };
}

export function getDisplayedAgreementIndicator(action = {}, agenda = null, index = 0) {
  const actionIndicator = action.agreement_indicator || deriveEducationAgreementIndicator(
    action.requester_entity_kind,
    action.requester_entity_nature,
    action.age_range
  );
  if (actionIndicator) return actionIndicator;
  const hasActionFields = Boolean(action.requester_entity_kind || action.requester_entity_nature || action.age_range);
  if (index === 0 && !hasActionFields) return deriveAgreementIndicatorFromAgenda(agenda);
  return "";
}

File: backend/apps/schedules/agreement_indicators.py
from django.db import models


class EducationAgreementIndicator(models.TextChoices):
    ESCOLINHA_NOTA_10 = "ESCOLINHA_NOTA_10", "Escolinha Nota 10"
    ESCOLA_NOTA_10 = "ESCOLA_NOTA_10", "Escola Nota 10"


class EducationActionAgeRange(models.TextChoices):
    AGE_05_10 = "AGE_05_10", "05 - 10 anos (ensino fundamental - anos iniciais)"
    AGE_11_14 = "AGE_11_14", "11 - 14 anos (ensino fundamental - anos finais)"
    AGE_15_17 = "AGE_15_17", "15 - 17 anos (ensino médio)"
    AGE_ADULT = "AGE_ADULT", "acima de 18 anos - Adultos"


class RequesterEntityKind(models.TextChoices):
    SCHOOL = "SCHOOL", "Instituição de Ensino"
    BUSINESS = "BUSINESS", "Empresa"
    EVENT_ORGANIZATION = "EVENT_ORGANIZATION", "Organização de Evento"
    MILITARY = "MILITARY", "Órgão Militar"
    PUBLIC = "PUBLIC", "Órgão Público"
    OTHER = "OTHER", "Outros"
    ADMINISTRATIVE = "ADMINISTRATIVE", "Demanda Administrativa"


class RequesterEntityNature(models.TextChoices):
    PUBLIC = "PUBLIC", "Pública"
    PRIVATE = "PRIVATE", "Privada"
    NOT_APPLICABLE = "NOT_APPLICABLE", "Não se aplica"


_SCHOOL_MARKERS = ("instituição de ensino", "instituicao de ensino", "escola", "colégio", "colegio")
_PUBLIC_MARKERS = (" público", " pública", "publico", "publica")
_PRIVATE_MARKERS = (" privado", " privada", "privado", "privada", "particular")

_AGENDA_AGE_RANGE_MAP = {
    "05 - 10 anos (ensino fundamental - anos iniciais)": EducationActionAgeRange.AGE_05_10,
    "11 - 14 anos (ensino fundamental - anos finais)": EducationActionAgeRange.AGE_11_14,
    "15 - 17 anos (ensino médio)": EducationActionAgeRange.AGE_15_17,
    "acima de 18 anos - Adultos": EducationActionAgeRange.AGE_ADULT,
}


def normalize_entity_type(requester_entity_type_str):
    text = str(requester_entity_type_str or "").strip()
    if not text:
        return None, None

    if text == "Demanda Administrativa" or text == RequesterEntityKind.ADMINISTRATIVE:
        return RequesterEntityKind.ADMINISTRATIVE, RequesterEntityNature.NOT_APPLICABLE

    if text == RequesterEntityKind.SCHOOL:
        return RequesterEntityKind.SCHOOL, None
    if text == RequesterEntityKind.BUSINESS:
        return RequesterEntityKind.BUSINESS, None
    if text == RequesterEntityKind.EVENT_ORGANIZATION:
        return RequesterEntityKind.EVENT_ORGANIZATION, None
    if text == RequesterEntityKind.MILITARY:
        return RequesterEntityKind.MILITARY, None
    if text == RequesterEntityKind.PUBLIC:
        return RequesterEntityKind.PUBLIC, None
    if text == RequesterEntityKind.OTHER:
        return RequesterEntityKind.OTHER, None

    if text == "2":
        return RequesterEntityKind.SCHOOL, None
    if text == "4":
        return RequesterEntityKind.BUSINESS, None
    if text == "1":
        return RequesterEntityKind.PUBLIC, None

    text_lower = text.lower()
    nature = None
    if any(marker in text_lower for marker in _PUBLIC_MARKERS):
        nature = RequesterEntityNature.PUBLIC
    elif any(marker in text_lower for marker in _PRIVATE_MARKERS):
        nature = RequesterEntityNature.PRIVATE

    if any(marker in text_lower for marker in _SCHOOL_MARKERS):
        return RequesterEntityKind.SCHOOL, nature

    if "empresa" in text_lower:
        return RequesterEntityKind.BUSINESS, nature

    if "organização de evento" in text_lower or "organizacao de evento" in text_lower:
        return RequesterEntityKind.EVENT_ORGANIZATION, nature

    if "militar" in text_lower:
        return RequesterEntityKind.MILITARY, nature

    if "órgão" in text_lower or "orgao" in text_lower:
        return RequesterEntityKind.PUBLIC, nature

    return None, None


def normalize_age_range(age_text):
    text = str(age_text or "").strip()
    if not text:
        return None

    if text in EducationActionAgeRange.values:
        return text

    return _AGENDA_AGE_RANGE_MAP.get(text)


def derive_education_agreement_indicator(kind, nature, age_range):
    internal_age = normalize_age_range(age_range)
    if (
        kind != RequesterEntityKind.SCHOOL
        or nature != RequesterEntityNature.PUBLIC
        or not internal_age
    ):
        return None

    if internal_age in (EducationActionAgeRange.AGE_05_10, EducationActionAgeRange.AGE_11_14):
        return EducationAgreementIndicator.ESCOLINHA_NOTA_10

    if internal_age in (EducationActionAgeRange.AGE_15_17, EducationActionAgeRange.AGE_ADULT):
        return EducationAgreementIndicator.ESCOLA_NOTA_10

    return None


def derive_from_agenda(agenda):
    if not agenda:
        return None
    req_type = getattr(agenda, "requester_entity_type", "") if hasattr(agenda, "requester_entity_type") else agenda.get("requester_entity_type", "")
    age_ranges = getattr(agenda, "age_ranges", "") if hasattr(agenda, "age_ranges") else agenda.get("age_ranges", "")
    kind, nature = normalize_entity_type(req_type)
    return derive_education_agreement_indicator(kind, nature, age_ranges)
